raft/src/main/java/org/apache/kafka/raft/internals/EpochElection.java

```diff
@@ -0,0 +1,216 @@
+    }
+
+    /**
+     * Get the collection of voter states.
+     */
+    public Collection<VoterState> voterStates() {
+        return Collections.unmodifiableCollection(voterStates.values());
+    }
+
+    /**
+     * Check whether we have received enough votes to conclude the election and become leader.
+     *
+     * @return true if at least a majority of nodes have granted the vote
+     */
+    public boolean isVoteGranted() {
+        return numGranted() >= majoritySize();
+    }
+
+    /**
+     * Check if we have received enough rejections that it is no longer possible to reach a
+     * majority of grants.
+     *
+     * @return true if the vote is rejected, false if the vote is already or can still be granted
+     */
+    public boolean isVoteRejected() {
+        return numGranted() + numUnrecorded() < majoritySize();
+    }
+
+    /**
+     * Get the set of voters which have not been counted as granted or rejected yet.
+     *
+     * @return The set of unrecorded voters
+     */
+    public Set<ReplicaKey> unrecordedVoters() {
+        return votersOfState(VoterState.State.UNRECORDED).collect(Collectors.toSet());
+    }
+
+    /**
+     * Get the set of voters that have granted our vote requests.
+     *
+     * @return The set of granting voters, which should always contain the localId
+     */
+    public Set<Integer> grantingVoters() {
+        return votersOfState(VoterState.State.GRANTED).map(ReplicaKey::id).collect(Collectors.toSet());
+    }
+
+    /**
+     * Get the set of voters that have rejected our candidacy.
+     *
+     * @return The set of rejecting voters
+     */
+    public Set<Integer> rejectingVoters() {
+        return votersOfState(VoterState.State.REJECTED).map(ReplicaKey::id).collect(Collectors.toSet());
+    }
+
+    private VoterState getVoterStateOrThrow(int voterId) {
+        VoterState voterState = voterStates.get(voterId);
+        if (voterState == null) {
+            throw new IllegalArgumentException("Attempt to access voter state of non-voter " + voterId);
+        }
+        return voterState;
+    }
+
+    private Stream<ReplicaKey> votersOfState(VoterState.State state) {
+        return voterStates
+            .values()
+            .stream()
+            .filter(voterState -> voterState.state().equals(state))
+            .map(VoterState::replicaKey);
+    }
+
+    private long numGranted() {
+        return votersOfState(VoterState.State.GRANTED).count();
+    }
+
+    private long numUnrecorded() {
+        return votersOfState(VoterState.State.UNRECORDED).count();
+    }
+
+    private int majoritySize() {
+        return voterStates.size() / 2 + 1;
+    }
+
+    @Override
+    public String toString() {
+        return String.format(
+            "EpochElection(%s)",
+            voterStates.values().stream()
+                .map(VoterState::toString)
+                .collect(
+                    Collectors.joining(", "))
+        );
```

Comment:
Why not just print the map?
```java
        return String.format(
            "EpochElection(voterStates=%s)",
            voterStates
        );
```

Reply:
it felt redundant to print the keys given that the replica ids are also contained in the values. since this is would only be used for debugging though, I'll take your suggestion and just print the entire map